src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -327,20 +354,57 @@ private static object GetHostEntryOrAddressesCore(IPAddress address, bool justAd
 
             // Do a reverse lookup to get the host name.
-            string? name = NameResolutionPal.TryGetNameInfo(address, out SocketError errorCode, out int nativeErrorCode);
-            if (errorCode != SocketError.Success)
+            ValueStopwatch stopwatch = NameResolutionTelemetry.Log.BeforeResolution(address);
+
+            SocketError errorCode;
+            string? name;
+            try
+            {
+                name = NameResolutionPal.TryGetNameInfo(address, out errorCode, out int nativeErrorCode);
+                if (errorCode != SocketError.Success)
+                {
+                    if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(address, $"{address} DNS lookup failed with {errorCode}");
+                    throw SocketExceptionFactory.CreateSocketException(errorCode, nativeErrorCode);
+                }
+                Debug.Assert(name != null);
+            }
+            catch when (LogFailure(stopwatch))
             {
-                if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(address, $"{address} DNS lookup failed with {errorCode}");
-                throw SocketExceptionFactory.CreateSocketException(errorCode, nativeErrorCode);
+                Debug.Fail("LogFailure should return false");
+                throw;
             }
 
+            NameResolutionTelemetry.Log.AfterResolution(stopwatch, successful: true);
+
             // Do the forward lookup to get the IPs for that host name
-            errorCode = NameResolutionPal.TryGetAddrInfo(name!, justAddresses, out string? hostName, out string[] aliases, out IPAddress[] addresses, out nativeErrorCode);
+            stopwatch = NameResolutionTelemetry.Log.BeforeResolution(name);
 
-            if (errorCode != SocketError.Success)
+            object result;
+            try
             {
-                if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(address, $"forward lookup for '{name}' failed with {errorCode}");
+                errorCode = NameResolutionPal.TryGetAddrInfo(name, justAddresses, out string? hostName, out string[] aliases, out IPAddress[] addresses, out int nativeErrorCode);
+
+                if (errorCode != SocketError.Success)
+                {
+                    if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(address, $"forward lookup for '{name}' failed with {errorCode}");
+                }
+
+                result = justAddresses ?
+                    (object)addresses :
+                    new IPHostEntry
+                    {
+                        HostName = hostName!,
+                        Aliases = aliases,
+                        AddressList = addresses
+                    };
             }
+            catch when (LogFailure(stopwatch))
+            {
+                Debug.Fail("LogFailure should return false");
+                throw;
+            }
+
+            NameResolutionTelemetry.Log.AfterResolution(stopwatch, successful: true);
```

Comment:
Please, check if the logging is enabled here at the call site to allow linker trim the call when appropriate. [Take a look at my PR.](https://github.com/dotnet/runtime/blob/abd9c9d468a32945582f2e61cf93d5d363d4efca/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L2326)